### DISCUSSION TITLE
Deduplicating types in a list

### DISCUSSION THREAD

**Discussion by spayne-pgh** [brunocodutra/metal#151] [Q&A] [answered]
We're looking at using Metal to replace an in-house type list implementation and one of the features that we use is to deduplicate types in a list. Looking at `boost::mp11`, there's a dedicated [mp_unique](https://www.boost.org/doc/libs/develop/libs/mp11/doc/html/mp11.html#mp_uniquel) template that will do that out-of-the box. But, we'd like to avoid boost if we could, just because of the size. Is there a ready-made solution for deduplicating types in a `metal::list` or would we have to roll our own?

**Comment by brunocodutra** [comment, marked as answer]
Hey @spayne-pgh, glad to hear you're considering Metal. An implementation of `unique` is not currently available out of the box, but can be easily implemented out of the primitives Metal offers, in fact it's a short one-liner:

```.cpp
template<class seq>
using unique = accumulate<bind<lambda<if_>, lambda<contains>, _1, lambda<append>>, list<>, seq>;
```

You can try it [live on compiler explorer](https://godbolt.org/z/5xKPfeb6T), is that what you were looking for?

**Reply by spayne-pgh** [reply]
Nice, that's an excellent one-liner. Thanks for the quick reply.
